### PR TITLE
Deterministic address contract factory implementation

### PR DESCRIPTION
**Pull request by z-j-lin**
## Scope

contract registry
## Why?

implementation of deterministic address contract deployment 

## Todos
Review and final clean up if possible

### CODE REVIEW

**Review by ricardopinto — CHANGES_REQUESTED** [review]
There's a respectable amount of assembly code here but only a handful unit tests. Given how central these changes are to the bridge, it would be best to extend the test suite and ensure greater coverage.

File: src/notes/code.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+pbc = "606061002e6053818101916020833d393d5182193002553639593060050152380310156029575936f35b3d3dfdfe363d3d377f000000000000000000000000000000000000000000000000000000000000000038198102903314604b573d903d903d9036903d90545af43d82803e156047573d90f35b3d90fd5b3d519055363df3"
```
Comment (vtleonardo) [inline src/notes/code.txt:1]: What is this byte string?

File: src/MadnetFactory.sol [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+            
+            contractAddr := create(0, basePtr, sub(ptr, basePtr))
+        }
+        codeSizeZeroRevert(uint160(contractAddr) != 0);
+        emit DeployedRaw(contractAddr);
+        return contractAddr;        
+    }
+
+    /**  
+    * @dev deployCreate2 deploys a contract from the factory address using create
```
Comment (vtleonardo) [inline src/MadnetFactory.sol:278]:
```suggestion
    * @dev deployCreate2 deploys a contract from the factory address using create2
```

File: src/MadnetFactory.sol [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+            let basePtr := mload(0x40)
+            let ptr := basePtr
+            
+            //copies the initialization code of the implementation contract
+            calldatacopy(ptr, _deployCode.offset, _deployCode.length)
+
+            //Move the ptr to the end of the code in memory
+            ptr := add(ptr, _deployCode.length)
+            
+            contractAddr := create(0, basePtr, sub(ptr, basePtr))
```
Comment (vtleonardo) [inline src/MadnetFactory.sol:270]: Are _deployCode.length and sub(ptr, basePtr)? If yes, maybe using _deployCode.length would be a little bit more clear.

File: src/MadnetFactory.sol [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+        }
+        codeSizeZeroRevert(uint160(contractAddr) != 0);
+        //add the salt to the list of contract names
+        contracts_.push(_salt);
+        emit Deployed(_salt, contractAddr);
+        return contractAddr;
+    }
+
+    /**  
+    * @dev requireAuth reverts if false and returns csize0 error message
```
Comment (vtleonardo) [inline src/MadnetFactory.sol:171]:
```suggestion
    * @dev requireAuth reverts if false and returns unauthorized error message 
```

File: src/MadnetFactory.sol [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+    }
+    /**  
+    * @dev deploy uses create2 to deploy a metamorphic contract to a deterministic location 
+    * once deployed the metamorphic contract will static call into the factory fall back to retrieve 
+    * the address of the most recent template deployment
+    * the metamorphic contract will then make a delegate call into the template contract to triger the deploycode which will 
+    * return the runtime code.
+    * after  deployment is done, if Init call data 
+    * @param _implementation the implementation contract address, to copy runtime code from.
+    if the 0 address is used, the last addressed stored on the global state is used
```
Comment (vtleonardo) [inline src/MadnetFactory.sol:139]:
```suggestion
    if the 0 address is used, the last address stored on the global state is used    
```
?

File: src/MadnetFactory.sol [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+    constructor(address bar, bytes memory) {
+        owner = bar;
+    }
+
+    fallback() external {
+        owner = msg.sender;
+    }
+}
+
+interface MadnetFacInterface {
```
Comment (vtleonardo) [inline src/MadnetFactory.sol:20]: Maybe: IMadnetFactory ?

File: src/MadnetFactory.sol [hunk 7]
```diff
@@ -0,0 +1,397 @@
+// SPDX-License-Identifier: MIT-open-group
+pragma solidity  ^0.8.11;
+import "../lib/utils/DeterministicAddress.sol";
+
+interface ifoo {
```
Comment (vtleonardo) [inline src/MadnetFactory.sol:5]: Are these dummy contracts being used for something? Maybe moving them to the test folder would be better, or put them in a mock folder?

**Review by vtleonardo — COMMENTED** [review]
Can we also move all the factory files to an `src/factory` folder for better organization?

File: lib/utils/DeterministicAddress.sol [hunk 8]
```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: MIT-open-group
+pragma solidity ^0.8.0;
+
+abstract contract DeterministicAddress {
+    // byte code for the switcheroo contract
+    // 6020363636335afa1536363636515af43d36363e3d36f3
```
Comment (vtleonardo) [inline lib/utils/DeterministicAddress.sol:6]: Only `6020363636335afa1536363636515af43d36363e3d36f3`is the whole byte code?

File: src/MadnetFactory.sol [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+    /**  
+    * @dev getNumContracts retrieves the address of the contract specified by its 32 byte salt derived from its name    
+    */
+    function getNumContracts() external view returns (uint256) {
+        return contracts_.length;
+    }
+
+    //returns implementation contracts address
+    /**  
+    * @dev deployTemplate deploys intermediate contract with the implementation contracts deploy code as its runtime code
```
Comment (vtleonardo) [inline src/MadnetFactory.sol:205]:
```suggestion
    * @dev deployTemplate deploys intermediate contract with the implementation contracts deployment code as its runtime code
```

File: src/MadnetFactory.sol [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+            ptr := add(ptr, 0x0b)
+            // modify runtime to contain the tail jump operation
+            // <codesize other> 56 5b 
+            // account for array offset
+            mstore8(ptr, 0x56)
+            ptr := add(ptr, 0x01)
+            mstore8(ptr, 0x5b)
+            ptr := add(ptr, 0x01)
+            //copies the initialization code of the implementation contract
+            calldatacopy(ptr, add(0x05, _deployCode.offset), sub(_deployCode.length, 0x05))
```
Comment (vtleonardo) [inline src/MadnetFactory.sol:227]: Any special reason why we skip the 0x05 first bytes?

File: src/MadnetFactory.sol [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+            ptr := sub(ptr, 0x20)
+            mstore(ptr, 0x28)
+            ptr := add(ptr, 0x20)
+            // finish the code with the terminate sequence  
+            // 5b 60 80 60 40 52 60 05 33 73 <factory> 14 15 57 33 ff fe
+            mstore8(ptr, 0x5b)
+            ptr := add(ptr, 0x01)
+            calldatacopy(ptr, _deployCode.offset, 0x05)
+            ptr := add(ptr, 0x05)
+            mstore(ptr, or(or(shl(240, 0x3373), shl(80, address())), shl(48, 0x14156004)))
```
Comment (vtleonardo) [inline src/MadnetFactory.sol:242]: Any meaning for these magic numbers?

**Pull request closed without merging**